File: 210-course-schedule-ii/course-schedule-ii.rb
# @param {Integer} num_courses
# @param {Integer[][]} prerequisites
# @return {Integer[]}
def find_order(num_courses, prerequisites)
  graph = Hash.new { |h, k| h[k] = [] }
  indegrees = [0] * num_courses
  prerequisites.each do |x, y|
    graph[y] << x
    indegrees[x] += 1
  end
  queue = []
  order = []
  indegrees.each_with_index { |el, index| queue << index if el == 0 }
  while !queue.empty?
    el = queue.shift
    order  << el
    graph[el].each do |node|
      if indegrees[node] != 0
        indegrees[node] -= 1
        queue << node if indegrees[node] == 0
      end
    end
  end
  order.size == num_courses ? order : []
end

File: 210-course-schedule-ii/course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        indegree = [0] * numCourses
        graph = defaultdict(list)
        
        for x, y in prerequisites:
            graph[y].append(x)
            indegree[x] += 1
        
        queue = deque([i for i in range(numCourses) if indegree[i] == 0])
        order = []
        while queue:
            node = queue.pop()
            order.append(node)
            for neighbor in graph[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)
        return order if len(order) == numCourses else []
            

        